### PR TITLE
[mlir][ArmSME] Remove ArmSMETypeConverter (and configure LLVM one instead)

### PR DESCRIPTION
**Pull request by MacDue**
This patch removes the ArmSMETypeConverter, and instead updates `populateArmSMEToLLVMConversionPatterns()` to add an ArmSME vector type conversion to the existing LLVMTypeConverter. This makes it easier to add these patterns to an existing `-to-llvm` lowering pass.

### CODE REVIEW

File: mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp [hunk 1]
```diff
@@ -574,10 +573,17 @@ void mlir::configureArmSMEToLLVMConversionLegality(ConversionTarget &target) {
       arm_sme::aarch64_sme_mopa>();
   target.addLegalDialect<arith::ArithDialect>();
   target.addLegalOp<UnrealizedConversionCastOp>();
+  typeConverter.addConversion([&](VectorType type) -> std::optional<Type> {
+    // There's no LLVM type for SME tiles, but after lowering to intrinsics all
+    // SME vector types should be eliminated.
+    if (arm_sme::isValidSMETileVectorType(type))
+      return type;
+    return std::nullopt;
+  });
```
Comment (c-rhodes) [inline mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:582]: nit: move to `populateArmSMEToLLVMTypeConversion`?

**Review by c-rhodes — APPROVED** [review]
grepping on `typeConverter.addConversion` this seems consistent with other conversions. One minor suggestion, but otherwise LGTM, cheers.

File: mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp (hunk 1 above)
Comment (c-rhodes) [reply mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp]: I meant a separate populate method that's called from configure (based on what I saw in SPIR-V which looked clean), but it's not so important :)

Comment (MacDue) [reply mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp]: Oh, I was looking at OpenMPToLLVM, which does it like this :)

Comment (c-rhodes) [reply mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp]: no worries, that's fine :)

**Comment by banach-space** [comment]
This will update the global `LLVMTypeConverter` used during the compilation, right? And so it will update it for all its consumers (even non-SME)? But that should be OK, because it will only happen when lowering to SME?

I just want to make sure that we retain the original protection that we gained when splitting this into a dedicated type converter.

**Comment by MacDue** [comment]
> This will update the global `LLVMTypeConverter` used during the compilation, right? And so it will update it for all its consumers (even non-SME)? But that should be OK, because it will only happen when lowering to SME?
> 
> I just want to make sure that we retain the original protection that we gained when splitting this into a dedicated type converter.

It's not global state, it's updating the single instance of the `LLVMTypeConveter` that's passed to it. Within MLIR core this is a NFC (since we construct a `LLVMTypeConveter` in our `-convert-arm-sme-to-llvm` pass). 

Note that the `ArmSMETypeConverter` did not override any methods, it just called `addConversion` within its constructor. Constructing a new `LLVMTypeConverter`, and then calling `addConversion()` on it is equivalent to that (without an extra class). 

Also, note that adding type conversions needed for patterns is a common thing done within other dialects (such as OpenMP and 
SPIR-V).

**Comment by banach-space** [comment]
> > This will update the global `LLVMTypeConverter` used during the compilation, right? And so it will update it for all its consumers (even non-SME)? But that should be OK, because it will only happen when lowering to SME?
> > I just want to make sure that we retain the original protection that we gained when splitting this into a dedicated type converter.
> 
> It's not global state, it's updating the single instance of the `LLVMTypeConveter` that's passed to it. Within MLIR core this is a NFC (since we construct a `LLVMTypeConveter` in our `-convert-arm-sme-to-llvm` pass).
> 
> Note that the `ArmSMETypeConverter` did not override any methods, it just called `addConversion` within its constructor. Constructing a new `LLVMTypeConverter`, and then calling `addConversion()` on it is equivalent to that (without an extra class).
> 
> Also, note that adding type conversions needed for patterns is a common thing done within other dialects (such as OpenMP and SPIR-V).

A slightly different question then - is there a test that shows that type conversion fails for vectors scalable in 2 dims? Unless lowering for SME? I think that that would be quite helpful.

**Comment by banach-space** [comment]
> Nothing happens, no patterns match, so no type conversions run, and because it's a partial conversion it also does not error.
> 
> See: https://godbolt.org/z/TTG9cGrrj

Argh, I missed that! Yeah, that makes sense, thanks for the discussion 🙏🏻

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

**Comment by MacDue** [comment]
> > Nothing happens, no patterns match, so no type conversions run, and because it's a partial conversion it also does not error.
> > See: https://godbolt.org/z/TTG9cGrrj
> 
> Argh, I missed that! Yeah, that makes sense, thanks for the discussion 🙏🏻

I've come up with a test now :)
I just had to write in in C++ rather than as an IR test, see:
https://github.com/llvm/llvm-project/blob/5e83afd0171c212a4d5c15db8e263650ac25e0f0/mlir/unittests/Dialect/ArmSME/TileTypeConversionTest.cpp

**Pull request merged**